prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -0,0 +1,336 @@
+     *
+     * Note:  There is a strong OO argument that this list should be broken out into each individual child Report.kt.
+     * (That is, every report should know its own parents!)
+     * However, its here because there are Functions that do not create Report.kt objects.  For example, Send.
+     * In addition, in-memory, reports get copied many times, with lots of parent-child relationships
+     * that are error-prone to track.  Hiding the lineage data here helps ensure correctness and hide complexity.
+     *
+     * todo Note that this does not work for command line.   Is that a problem?
+     * todo this is redundant with `Report.sources`.   Merge these together; eliminate one of them.
+     *
+     */
+    private val reportLineages = mutableListOf<ReportLineage>()
+
+    constructor(actionStr: String, context: ExecutionContext? = null) {
+        try {
+            action.actionName = TaskAction.valueOf(actionStr)
+            this.context = context
+        } catch (e: IllegalArgumentException) {
+            error("Unknown action $actionStr")
+        }
+    }
+
+    fun trackActionParams(request: HttpRequestMessage<String?>) {
+        val factory = JsonFactory()
+        val outStream = ByteArrayOutputStream()
+        factory.createGenerator(outStream).use {
+            it.useDefaultPrettyPrinter()
+            it.writeStartObject()
+            it.writeStringField("method", request.httpMethod.toString())
+            it.writeObjectFieldStart("Headers")
+            // remove secrets
+            request.headers.filter { !it.key.contains("key") }.forEach { (key, value) ->
+                it.writeStringField(key, value)
+            }
+            it.writeEndObject()
+            it.writeObjectFieldStart("QueryParameters")
+            // remove secrets
+            request.queryParameters.filter { !it.key.contains("code") }.forEach { (key, value) ->
+                it.writeStringField(key, value)
+            }
+            it.writeEndObject()
+            it.writeEndObject()
+        }
+        // truncate if needed
+        action.actionParams = outStream.toString().chunked(512)[0]
+    }
+
+    fun trackActionResult(actionResult: String) {
+        // TODO should be able to get this max size from the jooq-generated code.
+        // chop the string down to 2048 if needed.
+        action.actionResult = actionResult.chunked(2048)[0]
+    }
+
+    fun trackActionResult(httpResponseMessage: HttpResponseMessage) {
+        trackActionResult(httpResponseMessage.status.toString() + ":\n" + httpResponseMessage.body.toString())
+    }
+
+    /**
+     * No report can be tracked twice, either as an input or output.   Prevents loops and other shenanigans.
+     */
+    private fun reportAlreadyTracked(id: UUID): Boolean {
+        return reportsReceived.containsKey(id) ||
+            reportsIn.containsKey(id) ||
+            reportsOut.containsKey(id)
+    }
```

Comment:
A nice sanity check called every time a new report is tracked by this ActionHistory obj.